kyo-prelude/shared/src/main/scala/kyo/Batch.scala

```diff
@@ -0,0 +1,93 @@
+package kyo
+
+import kyo.Tag
+import kyo.kernel.*
+
+sealed trait Batch[+S] extends ArrowEffect[Batch.Op[*, S], Id]
+
+object Batch:
+
+    enum Op[V, -S]:
+
+        case Eval[A](seq: Seq[A]) extends Op[A, Any]
+
+        case Call[A, B, S](
+            v: A,
+            source: Seq[A] => Seq[B] < S,
+            frame: Frame
+        ) extends Op[B, S]
+    end Op
+
+    private inline def erasedTag[S]: Tag[Batch[S]] = Tag[Batch[Any]].asInstanceOf[Tag[Batch[S]]]
+
+    inline def source[A, B, S](f: Seq[A] => Seq[B] < S)(using inline frame: Frame): A => B < Batch[S] =
+        (v: A) => ArrowEffect.suspend[B](erasedTag[S], Op.Call(v, f, frame))
+
+    inline def eval[A](seq: Seq[A])(using inline frame: Frame): A < Batch[Any] =
+        ArrowEffect.suspend[A](erasedTag[Any], Op.Eval(seq))
+
+    def run[A: Flat, S, S2](v: A < (Batch[S] & S2))(using Frame): Seq[A] < (S & S2) =
+
+        case class Cont(op: Op[?, S], cont: Any => (Cont | A) < (Batch[S] & S2))
+        def runCont(v: (Cont | A) < (Batch[S] & S2)): (Cont | A) < S2 =
+            // TODO workaround, Flat macro isn't inferring correctly with nested classes
+            import Flat.unsafe.bypass
+            ArrowEffect.handle(erasedTag[S], v) {
+                [C] => (input, cont) => Cont(input, v => cont(v.asInstanceOf[C]))
+            }
+        end runCont
+
+        case class Expanded(v: Any, source: Seq[Any] => Seq[Any] < S, cont: Any => (Cont | A) < (Batch[S] & S2), frame: Frame)
```

Comment:
This code is not optimized since it uses several Scala collection operations. I think it's ok for an initial version so we can validate the effect and its functionality.